tests/test_unit/test_cdf_tk/test_client/test_extended_functions.py

```diff
@@ -0,0 +1,113 @@
+        assert result.external_id == "test_function"
+        assert console.print.call_count == global_config.max_retries - 1
+        assert console.print.call_args.args[1] == "Rate limit exceeded. Retrying after 42 seconds."
+
+    def test_create_function_429_exceed_max_retries(self, toolkit_config: ToolkitClientConfig) -> None:
+        client = ToolkitClient(config=toolkit_config, enable_set_pending_ids=True)
+        url = f"{toolkit_config.base_url}/api/v1/projects/test-project/functions"
+        fun = FunctionWrite(name="test_function", file_id=123, external_id="test_function")
+        console = MagicMock(spec=Console)
+        with responses.RequestsMock() as rsps, patch(f"{ExtendedFunctionsAPI.__module__}.time.sleep"):
+            for i in range(global_config.max_retries + 1):
+                rsps.add(
+                    responses.POST, url, status=429, json={"error": "Too many requests"}, headers={"Retry-After": "42"}
+                )
+            with pytest.raises(CogniteAPIError) as exc_info:
+                client.functions.create_with_429_retry(fun, console=console)
+        assert console.print.call_count == global_config.max_retries
+        assert exc_info.value.message == "Too many requests"
+        assert exc_info.value.code == 429
+
+    def test_create_function_429_invalid_retry_after(self, toolkit_config: ToolkitClientConfig) -> None:
+        client = ToolkitClient(config=toolkit_config, enable_set_pending_ids=True)
+        url = f"{toolkit_config.base_url}/api/v1/projects/test-project/functions"
+        fun = FunctionWrite(name="test_function", file_id=123, external_id="test_function")
+        console = MagicMock(spec=Console)
+        with responses.RequestsMock() as rsps, patch(f"{ExtendedFunctionsAPI.__module__}.time.sleep"):
+            rsps.add(
+                responses.POST, url, status=429, json={"error": "Too many requests"}, headers={"Retry-After": "invalid"}
+            )
+            with pytest.raises(CogniteAPIError) as exc_info:
+                client.functions.create_with_429_retry(fun, console=console)
+        assert console.print.call_count == global_config.max_retries
+        assert "Rate limit exceeded. Retrying after 60 seconds." in console.print.call_args.args[1]
+        assert exc_info.value.message == "Too many requests"
+
+    def test_create_function_401(self, toolkit_config: ToolkitClientConfig) -> None:
+        client = ToolkitClient(config=toolkit_config, enable_set_pending_ids=True)
+        url = f"{toolkit_config.base_url}/api/v1/projects/test-project/functions"
+        fun = FunctionWrite(name="test_function", file_id=123, external_id="test_function")
+        with responses.RequestsMock() as rsps:
+            rsps.add(responses.POST, url, status=401, json={"error": "Unauthorized"})
+            with pytest.raises(CogniteProjectAccessError):
+                client.functions.create_with_429_retry(fun)
```

Comment:
I don't think this asserts that it did not retry on 401, as the request mock will continue to respond with 401 indefinitely.

Reply:
Added extra assert.